internal/pkg/v2/README.md

```diff
@@ -0,0 +1,275 @@
+
+
+
+### Use-cases
+
+Use-cases live in the application layer. 
+
+A single use-case can handle one or more version, type, and action variations.
+
+A use-case typically has its own specific request DTO (for input) and response DTO (for output).
+
+A use-case is implemented following the [Gang of Four Command design pattern](https://en.wikipedia.org/wiki/Command_pattern) and implements the `Routable` contract.
+
+
+
+### DTO Definitions
+
+Request and Response DTOs are defined in the application layer.  To facilitate acceptance testing (see elsewhere in this document), all DTO variations for all minor releases are maintained in a versioned directory structure (see `internal/pkg/v2/application/dto`).
+
+
+
+### DTO Validation
+
+DTO validation lives in the application layer; specifically in `internal/pkg/v2/application/validator`.
+
+Validation is an orthogonal concern and occurs outside of use-case implementation.
+
+Validators are applied to use-cases during wire-up.  Validators can be layered.
+
+Validation shares implementation with middleware; the underlying implementation to curry/wrap `application.Executable` behavior is common for both.
+
+
+
+### Router
+
+The router lives in the user interface layer.  
+
+It maps version, type, and action to a `Routable` contract implementation.
+
+
+
+### Controllers
+
+Controllers live in the user interface layer and implement the `Controller` contract.
+
+They bridge the user interface (in the current implementation, this is an endpoint invocation) to a specific use-case implementation.
+
+
+
+### Handlers
+
+Two generic handler implementations are provided -- one for all use-case endpoint invocations and one for batch endpoint invocations.  
+
+These generic implementations provide generic marshalling and unmarshalling of JSON to and from Golang structures.  They use the Router to determine what Use-case implementation to delegate to.
+
+
+
+### Middleware
+
+Middleware lives in the user interface layer.
+
+It typically contains behavior executed for each use-case request.  It supports both pre- and post-use-case processing.  It can modify the request (as it heads towards the use-case for processing) and/or the response (as it heads towards the user after use-case processing has completed).
+
+Middleware is optional and can be selectively enabled.  It can be added to all use-cases for universal orthogonal concerns (the debugging middleware provides a good example of this).  It can also be selectively applied to individual use-cases.
+
+Middleware is extensible and can be layered.  You can apply different middleware implementations to the same use-case; each middleware is implemented and executed in isolation from every other middleware.
+
+Middleware shares implementation with DTO validation; the underlying implementation to curry/wrap `application.Executable` behavior is common for both.
```

Comment:
Not familiar with the term `curry/wrap`. What does this mean?

Reply:
> In mathematics and computer science, currying is the technique of translating the evaluation of a function that takes multiple arguments into evaluating a sequence of functions, each with a single argument. For example, a function that takes two arguments, one from _X_ and one from _Y_, and produces outputs in _Z_, by currying is translated into a function that takes a single argument from _X_ and produces as outputs functions from _Y_ to _Z_.

https://en.wikipedia.org/wiki/Currying

Reply:
I am very familiar with the technique, but never heard it referred to as Currying.